File: exercises/python/ch01/train_unigram.py
#!/usr/bin/env python
# coding=utf-8

from __future__ import unicode_literals, print_function, division
from itertools import chain
from collections import Counter
import sys, io, os

__train_input     = "../../../test/01-train-input.txt"
__train_data_file = "../../../data/wiki-en-train.word"
__model_file      = "./model_file"

def report(counter):
  ur"""
  >>> counter = train_unigram(__train_input)
  >>> report(counter)
  >>> print(io.open(__model_file).read()) # doctest: +NORMALIZE_WHITESPACE
  </s> 0.25
  a 0.25
  b 0.25
  c 0.125
  d 0.125
  """
  total_count = sum(count for (word, count) in counter.iteritems())

  with io.open(__model_file, mode="w", encoding='utf8') as f:
    for (word, prob) in sorted((word, count / total_count) for (word, count) in counter.iteritems()):
      print("{0}\t{1}".format(word, prob), file=f)
      
def train_unigram(filename):
  counter = Counter()
  counter.update(word for line in io.open(filename) for word in chain(line.split(), ["</s>"])) 
  return counter
  
if __name__ == '__main__':
  filename = sys.argv[1] if len(sys.argv) >= 2 and os.path.isfile(sys.argv[1]) else __train_data_file
  print(filename)
  counter = train_unigram(filename)
  report(counter)
  

File: exercises/scala/ch01/TrainUnigram.scala
import scala.collection.mutable.{Map, ListBuffer}
import scala.collection.immutable.TreeMap

object TrainUnigram extends App {
  val counter = Map[String, Int]().withDefaultValue(0)
  
  val TrainInput     = "../../../test/01-train-input.txt"
  val TrainDataFile  = "../../../data/wiki-en-train.word"
  val ModelFile     = "./model_file"
  val filename = if (args.length >= 1 && new java.io.File(args(0)).exists) args(0) else TrainInput
  
  io.Source.fromFile(filename).getLines.toList.foreach {(line:String) =>
    val words = line.split("[ \t]+").to[ListBuffer] += "</s>"
    for (word <- words) {
      counter.update(word, counter(word) + 1)
    }
  }
  val total_count = counter.valuesIterator.sum
  // http://stackoverflow.com/a/3074156/1542390
  
  try {
    val outs = new java.io.PrintStream(new java.io.File(ModelFile))
      
    TreeMap(counter.toArray:_*).foreach { case (word:String, count:Int) =>
      Console.withOut(outs) { println(word + "\t" + count) }
    }
  } catch {
    case e:Exception => println("Exception")
  }
}